Add purchase modal at the end of the profile wizard

justinshreve opened the pull request:
Closes #3441.

This PR adds a modal at the end of the profile wizard that allows the user to confirm if they would like to purchase products now, or later (via the task list). I've tried to build this in a way that the modal can easily be used from the task as well.

It also fixes the intermittent bug we were seeing at the end of the theme step where the app would get stuck in a `updateProfileItems` loop.

One thing that I think needs further follow-up, is selecting Storefront or one of the other free themes still shows the dialog and redirects you to the cart (`master` also currently adds these items to the WC.com cart). I think we'll want to install the free themes from WordPress.org instead. cc @joshuatf.

### Screenshots

<img width="681" alt="Screen Shot 2019-12-18 at 3 15 09 PM" src="https://user-images.githubusercontent.com/689165/71120572-b1d67400-21aa-11ea-9fd2-1bfbd9d62164.png">

### Detailed test instructions:

* Enable the profile wizard
* Make a product type selection or theme selection that requires purchase
* Finish the theme step, and see the modal appear
* Walk through the profile wizard again and make selections that don't require purchase
* No modal should be displayed

joshuatf commented on client/dashboard/utils.js:
```diff
@@ -36,6 +36,35 @@ export function getCurrencyRegion( countryState ) {
 }
 
+/**
+ * Gets the product IDs for items selected for purchase.
```
Nice - thanks for moving this 👍

Small nitpick: might update this description to read:

```suggestion
 * Gets the product IDs for items based on the product types and theme selected in the onboarding profiler.
```

justinshreve replied:
Fixed in https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/d7762abe8ad6717ef3db01792f1b1fe9f3584906.

joshuatf commented on client/dashboard/profile-wizard/cart-modal.js:
```diff
@@ -0,0 +1,134 @@
+					'Would you like to purchase and install the following features now?',
+					'woocommerce-admin'
+				) }
+				onRequestClose={ () => this.props.onClose() }
```
Not a big deal and not sure if there's a way around this, but this will also close when clicking outside the modal; which I would expect to return to the theme selection if it was in the background.

justinshreve replied:
I moved some logic around, so now this opens in front of the theme screen. You can now close the modal and reopen it.

https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/d7762abe8ad6717ef3db01792f1b1fe9f3584906

joshuatf commented on client/dashboard/style.scss:
```diff
@@ -179,3 +179,44 @@
 	}
 }
+
+.components-modal__frame.woocommerce-cart-modal {
```
There a few minor differences in heading and p tag sizes between this and the design.

#### Current
<img width="604" alt="Screen Shot 2019-12-19 at 2 15 32 PM" src="https://user-images.githubusercontent.com/10561050/71149831-be220600-226a-11ea-9530-acd5c50958e1.png">


#### Design
<img width="575" alt="Screen Shot 2019-12-19 at 2 21 14 PM" src="https://user-images.githubusercontent.com/10561050/71149869-d5f98a00-226a-11ea-8283-b1a2ede17270.png">

justinshreve replied:
Fixed in https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/d7762abe8ad6717ef3db01792f1b1fe9f3584906.

joshuatf commented on client/dashboard/index.js:
```diff
@@ -2,93 +2,54 @@
 			return <ProfileWizard query={ query } />;
 		}
 
+		if ( isOnboardingEnabled() && showCartModal ) {
```
The only issue with separating the modal from the profile wizard here is the theme previews can't be shown in the background.  May not be super important though; /cc @jameskoster in case you have a strong opinion one way or the other.

#### Current
<img width="997" alt="Screen Shot 2019-12-19 at 2 17 54 PM" src="https://user-images.githubusercontent.com/10561050/71150019-338dd680-226b-11ea-8721-d8ff9bbeae41.png">


#### Design
<img width="647" alt="Screen Shot 2019-12-19 at 2 23 45 PM" src="https://user-images.githubusercontent.com/10561050/71150023-3688c700-226b-11ea-8ab8-8d60ebead744.png">

justinshreve replied:
I'm moving some things around to make this possible.

justinshreve replied:
I have to move some things around to make this work (namely mark the wizard as completed at a different point) but this now opens in front of the theme step.

https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/d7762abe8ad6717ef3db01792f1b1fe9f3584906

joshuatf commented on client/dashboard/profile-wizard/cart-modal.js:
```diff
@@ -0,0 +1,134 @@
+				return theme.id === productId;
+			} );
+
+			if ( themeInfo && themeInfo.slug !== 'Storefront' ) {
```
Not opposed to separating Storefront and installing from the wp.org repo, but if we do that we may also want to check if this is the only product ID.  Otherwise, we should probably Storefront in this list since it will be added via wccom.

On a similar note, would we want to install any free themes from wp.org or explicitly set any that should bypass wccom?  If the former, we could check against price here to allow any themes that are free to be installed without visiting wccom.

justinshreve replied:
I think we should probably install any of the free themes from .org instead of WooCommerce.com. It saves multiple steps for the user if they didn't pick any other products, and it's less confusing than sending them to a cart with a $0.00 product.

It looks like the other free themes (like Deli) are in the .org repo.

What do you think about this being done in a follow-up? I think a couple pieces will need to be done. I can back out the `Storefront` check here and add a todo / log an issue.

joshuatf replied:
I think that's a good plan.  Outlined an issue for it [here](https://github.com/woocommerce/woocommerce-admin/issues/3455).

joshuatf commented on client/dashboard/profile-wizard/cart-modal.js:
```diff
@@ -0,0 +1,134 @@
+						{ __( "I'll do it later", 'woocommerce-admin' ) }
+					</Button>
+
+					<Button isPrimary isDefault onClick={ () => this.redirectToCart() }>
```
Optional: might be a nice addition for slower connections if we add a loading state when this is clicked.

justinshreve replied:
Added in https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/d7762abe8ad6717ef3db01792f1b1fe9f3584906.

justinshreve commented:
Hey @joshuatf I addressed your feedback in https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/d7762abe8ad6717ef3db01792f1b1fe9f3584906 -- except for the free themes (which I replied to [here](https://github.com/woocommerce/woocommerce-admin/pull/3444#discussion_r359921124)).

justinshreve commented on client/dashboard/profile-wizard/cart-modal.js:
```diff
@@ -102,19 +107,19 @@ class CartModal extends Component {
 					</Button>
 
-					<Button isPrimary isDefault onClick={ () => this.redirectToCart() }>
+					<Button isPrimary isDefault isBusy={ isRedirecting } onClick={ () => this.onClickNow() }>
```
Renamed to `onClickPurchaseNow` in https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/c4f2866960f415f07f174de0897d379e715435de.

justinshreve commented on client/dashboard/profile-wizard/cart-modal.js:
```diff
@@ -102,19 +107,19 @@ class CartModal extends Component {
 
 				<div className="woocommerce-cart-modal__actions">
-					<Button isLink onClick={ () => this.props.onClose() }>
+					<Button isLink onClick={ () => this.props.onClickLater() }>
```
Renamed to `onClickPurchaseLater` and added a busy state in https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/c4f2866960f415f07f174de0897d379e715435de.

justinshreve commented on client/dashboard/profile-wizard/index.js:
```diff
@@ -25,17 +26,31 @@ import Start from './steps/start';
 		super( ...arguments );
+		this.state = {
+			showCartModal: false,
+			cartRedirect: false,
```
Renamed in https://github.com/woocommerce/woocommerce-admin/pull/3444/commits/c4f2866960f415f07f174de0897d379e715435de.
